### DISCUSSION TITLE
Usb host support

### DISCUSSION THREAD

**Discussion by Geoide** [Moddable-OpenSource/moddable#920] [Getting started]
Hello.

Is there some way to reading data from a wireless gamepad (logitech). I’m finding some way to use a usb host shield with esp32 and moddable but I’m new in this library and I’m not sure if there is support to this. 

Any help to start with it will be welcome. 

Thanks

**Comment by andycarle** [comment]
Hi @Geoide,

To the best of my recollection, no one has implemented USB host support in the Moddable SDK. It would certainly be doable, but I don't believe it has already been done.

That said, if one could get the communication with the USB host controller working (i.e. Serial communication or SPI or however the part you have in mind works), the USB gamepad protocol support on top of that would be fairly straightforward. It would resemble (but not quite be identical to) the BLE Central HID examples that we have in the Moddable SDK, such as [hid-keyboard](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/ble/hid-keyboard) and [hid-mouse](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/ble/hid-mouse). 

What USB host shield are you looking at? 

Also, if pairing with a BLE gamepad is an option (instead of using USB), there would be a much quicker path to getting that working.

**Comment by Geoide** [comment]
Hello @andycarle.

Thanks for your response. I’m thinking ins some approach I’ve seem in this page (based in MAX3421E chip):
https://www.hackster.io/usini/plug-any-usb-device-on-an-esp8266-e0ca8a

Perhaps if I follow the pins connection from this page I could get the communication but I’m not sure. 

My target is to reuse a old wireless gamepad which set the connection from a usb receiver.

**Reply by andycarle** [reply]
Connecting that shield and communicating with it from an ESP32 would certainly be possible using the Moddable SDK. But you would have to write a SPI-based driver to communicate with the controller, then build out the logic for parsing the USB messages, and then write the actual application logic for USB HID on top of that. 

All very doable! But definitely a bit of a project. If you decided to give it a shot, we could certainly give you some pointers along the way.

**Reply by Geoide** [reply]
I could try it but I’m not expert in this field. What would be the starting point exactly? Is there documentation?

**Reply by andycarle** [reply]
The starting point here would probably be to adapt an existing SPI driver in the Moddable SDK to work with this USB shield. Before about a year ago, all of our SPI drivers were implemented in C with JavaScript bindings to interact with scripts. But we now also support SPI transactions in script using our implementation of the [Ecma-419 standard](https://419.ecma-international.org/#-10-io-classes-serial-peripheral-interface-spi). 

The spec text is essentially the documentation for that at the moment. And we have a handful of examples that use SPI from script:

- [xpt2046](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/spi/xpt2046)
- [ili9341](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/spi/ili9341)
- [display-ili9341](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/spi/display-ili9341)

Those would be the right place to get going on trying to make this USB shield work. 

It might be straightforward to adapt existing Arduino or C code into a working driver in JavaScript. That said, I don't want to mislead you: this project isn't necessarily what I would recommend as a first step into working with hardware in JavaScript. It won't be as simple as just interacting with a basic sensor. And that alone can be challenging if you've not worked much with hardware and a datasheet.

I'm happy to help if you go for it! But don't be surprised if this project takes some frustrating twists and turns along the way. :)

**Comment by barbiani** [comment]
Could esp32 s3 have usb support from the stack?

**Reply by phoddie** [reply]
Yes, it could. It should be pretty straightforward to build JavaScript bindings on top of tinyusb. I'm not sure if the WebUSB API would make sense or not. In particular, WebUSB depends on Web Streams, which are pretty heavy. Maybe a lighter embedded JavaScript API could be a better place to start. As I recall, tinyusb has modules for various kinds of devices (serial and mass storage, for example). Those don't fit WebUSB but could be mapped in through another API.
